internal/instrument.go

```diff
@@ -0,0 +1,145 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package internal
+
+import (
+	"fmt"
+	"go/parser"
+	"go/token"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/dave/dst"
+	"github.com/dave/dst/decorator"
+)
+
+const (
+	TargetPkg      = "main"
+	TargetFunc     = "main"
+	TrampolineName = "Trampoline"
+	HookName       = "Hook"
+)
+
+func loadAst(filePath string) *dst.File {
+	name := filepath.Base(filePath)
+	fset := token.NewFileSet()
+	file, err := os.Open(filePath)
+	if err != nil {
+		panic(err)
+	}
+	astFile, err := parser.ParseFile(fset, name, file, parser.ParseComments)
+	if err != nil {
+		panic(err)
+	}
+	dec := decorator.NewDecorator(fset)
+	dstFile, err := dec.DecorateFile(astFile)
```

Comment:
It would be neat if we could avoid taking the dependency on `dave/dst` (it does a lot beyond what is necessary to do our deed here; and adds a non-negligible allocations load).

Reply:
I agree with that. Personally, I'm fan of depending only on standard library, considering very carefully third party dependencies.

Reply:
But without `dst`, how do we insert code at the correct position? As far as I know the `go/ast` package wasn't created with source manipulation as an intended use-case. Comments are stored by their byte offset instead of attached to nodes, so re-arranging nodes breaks the output and confuses the debugger.

Reply:
True, that there is a problem with comments and in consequence it might confuse the debugger. That's strong argument.

Reply:
So let's keep the status quo. If needed later, we can trim down to a minimal version of dst that doesn't rely on any packages outside the standard library.

Reply:
@y1yang0 you can modify the AST with `go/ast` in pretty much the same way that you do it with `dave/dst`; just by modifying nodes or inserting new ones, and then you use `go/printer` to turn it back into go source.

The main thing that `dave/dst` makes easier is managing comments (such as `//go:linkname` pragmas). The way you've set things up in this prototype actually only uses `//go:linkname` on the `sdk` side, which I reckon contains only synthetic nodes or code that we have laid out previously so we may not even need to deal with comment placement at all.

Reply:
@RomainMuller Well, we can do everything with AST using go/ast, that's for sure. The challenge lies in building a mapping between the input source code and the output. When modifying AST nodes, tracking the relationship between the original and transformed positions can be difficult. Tools like token.Pos and go/token help but may not cover all cases. Actually, I'm not sure whether dave/dst offers any advantages in this regard, as I don't have experience with it.

Reply:
@RomainMuller As @pdelewski  said, go/ast cannot handle spaces/comments correctly, generating incorrect line numbers/column numbers, which can confuse the debugger.

Reply:
I think it's time to merge it. We can revisit some decisions later if needed.